### DISCUSSION TITLE
Idea: add template metadata to provisioner as a data source

### DISCUSSION THREAD

**Discussion by johnstcn** [coder/coder#5589] [Coder]
In the docker-image-builds example template](https://github.com/coder/coder/tree/main/examples/templates/docker-image-builds#updating-images) template, there is a manual required step of updating the image tag in the Terraform code. Without this step, the Terraform provider doesn't rebuild the image.

What would make intuitive sense here is to provide a data source that exposes the template version identifier to the provisioner so that the manual step above becomes automated:

```
resource "docker_image" "coder_image" {
  [...]
  tag = ["coder-${var.docker_image}:${data.coder_template.version_friendly}"]
  [...]
}
```

There are some other use-cases I can also see this enabling though:
- With cloud infra, you could tag resources with the template and version that originally spawned them. This could then be consumed by cost attribution tools etc.
- Similarly with Kubernetes resources, deployments, volumes etc. could be labeled with the template and version.

I'd imagine the schema to look something like the following:
- `name` (string) The name of the template
- `id` (string) The UUID of the template
- `version_id` (string) The UUID of the active template version
- `version` (string) The SHA-like representation of the active template version ID
- `version_friendly` (string) The human-readable UUID (e.g. fantastic_narwhal7) of the active template version ID

Thoughts?

**Comment by mtojek** [comment]
> What would make intuitive sense here is to provide a data source that exposes the template version identifier to the provisioner so that the manual step above becomes automated:

This would improve the user experience! I'm curious if we need a manual override, so that users can easily publish different templates with different version tags. Let's say we know that a published revision `v1.2` is broken, so DevOps would like to temporarily replace it with Docker image SHA.

> Similarly with Kubernetes resources, deployments, volumes etc. could be labeled with the template and version.
:+1

> I'd imagine the schema to look something like the following:

Do you think that we could support a nested structure, so that it bundles all version properties together under `version` key?

**Reply by johnstcn** [reply]
> Do you think that we could support a nested structure, so that it bundles all version properties together under version key?

I don't see why not!

**Reply by johnstcn** [reply]
> I'm curious if we need a manual override, so that users can easily publish different templates with different version tags. 

Would this instead be a use-case for rolling back to a previous template version?

**Reply by mtojek** [reply]
Rolling-back to the previous template can help here if it doesn't mean that it will nuke the latest template version. Let's say we want to keep it for debugging purposes.

Maybe it's an edge case, but I remember that I struggled a bit with emergency fixes applied to Docker `:latest` image.

**Comment by ddsmith2-eprod** [comment]
+1

Our company is using Artifactory and migrating from Coder v1 to v2.  Being able to get tags dynamically is a feature we would really like to see as well.

**Reply by matifali** [reply]
You mean to get all tags for an image from the registry,,?

**Reply by ddsmith2-eprod** [reply]
Yes.  In Coder v1 when users click to create a new workspace, they get a screen where they can select an existing image source and then select a tag.  They have reported to me that the Git integration dynamically updates the tags from Artifactory.

The feedback I received after setting up a v2 cluster was that they want to deploy the images they use for Airflow into Coder by just selecting the image and a tag.  When we spoke with support, they mentioned this might be a new feature that will get added eventually.  

It would greatly speed up our developer's adoption of the new platform.  We may be able to accomplish this by modifying their current workflow.  I just have not had time to investigate it.

**Reply by matifali** [reply]
There is a related issue that is no more being worked on I guess. https://github.com/kreuzwerker/terraform-provider-docker/issues/204

**Reply by matifali** [reply]
Created a new request: https://github.com/kreuzwerker/terraform-provider-docker/issues/524

**Reply by matifali** [reply]
There is now a native data source to fetch tags for an image.
https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs/data-sources/registry_image_tags
